### PR TITLE
CORE-14343: Add tracing sample rate configuration

### PR DESCRIPTION
**Pull request by josephzunigadaly**
* Add new command line option `--trace-samples-per-second`
* Defaults to 1 sample per second if not set
* Allows specifying unsigned numeric values as samples per second
* Allows "unlimited" value to always sample

### CODE REVIEW

File: libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt [hunk 1]
```diff
@@ -24,8 +25,13 @@ import zipkin2.reporter.urlconnection.URLConnectionSender
 import java.util.Stack
 import java.util.concurrent.ExecutorService
 import javax.servlet.Filter
+
+sealed interface SampleRate
+object Unlimited : SampleRate
+data class PerSecond(val samplesPerSecond: Int) : SampleRate
```
Comment (vkolomeyko) [inline libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt]:
```suggestion
internal sealed interface SampleRate
private object Unlimited : SampleRate
private data class PerSecond(val samplesPerSecond: Int) : SampleRate
```

File: libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt [hunk 2]
```diff
@@ -24,8 +25,13 @@ import zipkin2.reporter.urlconnection.URLConnectionSender
...
+
 @Suppress("TooManyFunctions")
-class BraveTracingService(serviceName: String, zipkinHost: String) : TracingService {
+class BraveTracingService(serviceName: String, zipkinHost: String, samplesPerSecond: SampleRate) : TracingService {
```
Comment (vkolomeyko) [inline libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt]:
```suggestion
internal class BraveTracingService(serviceName: String, zipkinHost: String, samplesPerSecond: SampleRate) : TracingService {
```

File: libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt [hunk 3]
```diff
@@ -35,12 +41,17 @@ class BraveTracingService(serviceName: String, zipkinHost: String) : TracingServ
             .addScopeDecorator(MDCScopeDecorator.get())
             .build()
 
+        val sampler = when (samplesPerSecond) {
+            is PerSecond -> RateLimitingSampler.create(samplesPerSecond.samplesPerSecond)
+            is Unlimited -> Sampler.ALWAYS_SAMPLE
+        }
+
```
Comment (vkolomeyko) [inline libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt:57]: I would add an `INFO` log line informing how frequently we will be collecting tracing samples.

File: libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt (hunk 1 above)
Comment (josephzunigadaly) [reply libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt]: Made them all internal for access from `TraceThis.kt`

File: libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt (hunk 3 above)
Comment (josephzunigadaly) [reply libs/tracing/src/main/kotlin/net/corda/tracing/impl/BraveTracingService.kt:57]: Done

**Review by vkolomeyko — APPROVED** [review]
LGTM

**Review by fenryka — APPROVED** [review]
LGTM

**Comment by driessamyn** [comment]
Should this not be a dynamic config that can be set through the config API ?

(not suggesting we undo this change, but wondering it that needs to be a follow-up JIRA?)

**Comment by vkolomeyko** [comment]
> Should this not be a dynamic config that can be set through the config API ?
> 
> (not suggesting we undo this change, but wondering it that needs to be a follow-up JIRA?)

IMO that would be un-necessary and overcomplicated as every worker would need to listen to config update and change tracing sampling in response.
If C5 cluster operator wants to change sampling rate, then can always modify value in Helm chart and update their deployment.

**Pull request merged**